File: 42.py
def trap(height):
    """
    :type height: List[int]
    :rtype: int
    """
    left = 0
    right = len(height) - 1
    minheight = 0
    result = 0
    while(left < right):
        l_v = height[left]
        r_v = height[right]
        if(l_v <= r_v):
            if l_v < minheight:
                result = result + minheight - l_v
            else:
                minheight = l_v
            left = left + 1
        else:
            if r_v < minheight:
                result = result + minheight - r_v
            else:
                minheight = r_v
            right = right - 1
    return result

print(trap([0,1,0,2,1,0,1,3,2,1,2,1]))
#print(trap([1]))

File: 42.trapping-rain-water/42.js
/*
 * @lc app=leetcode id=42 lang=javascript
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    var stack = [];
    var current = 0;
    var result = 0;
    for(var current = 0; current < height.length; current++){
      while(!isEmpty(stack) && (height[current] > height[getTop(stack)])){
        var top = getTop(stack);
        stack.pop();
        if(isEmpty(stack)){
          break;
        }
        var distance = current - getTop(stack) - 1;
        var bounded_height = Math.min(height[current], height[getTop(stack)]) - height[top];
        result += bounded_height * distance;
      }
      stack.push(current);
    }

    return result;

    function isEmpty(stack){
      return stack.length == 0;
    }
    function getTop(stack){
      return stack[stack.length - 1];
    }
      
};
// @lc code=end

